### ISSUE TITLE
New resid 8580 filter status?

### ISSUE THREAD

**Issue by vitamin-caig** [libsidplayfp/libsidplayfp#141]
Hello!

Looks like I came across some strange bug in new 8580 filter for resid builder (not resid-fp!). Playback for some tracks is broken (raw noise or corrupted sound). Moreover, result may vary between different starts for the same track from the same binary. Valgrind gives nothing - there're no any uninitialized memory read or something similar.
WavewormGenerator receives the same control values, so it's not an emulation issue. After switching to old filter problem disappears (at least for mingw64 build, looking forward for another platforms test results.

Looks like memory corruption? Is new filter code is ready to use?

**Comment by drfiemost** [comment]
Hi! Yeah, the code is production ready and currently used in VICE. It is more or less on par with residfp so if something sounds off it's a bug. Which tunes sound bad?

**Comment by vitamin-caig** [comment]
Robocop3.sid from Jeroen Tel - exactly first track sometimes sounds bad (only noise or slightly distorted sound).
Love_is_a_shield.sid - only some sounds available sometime.

**Comment by vitamin-caig** [comment]
Also report from another user: https://bitbucket.org/zxtune/zxtune/issues/260/android-build-r5070-breaks-sid-playback

**Comment by drfiemost** [comment]
> Also report from another user: https://bitbucket.org/zxtune/zxtune/issues/260/android-build-r5070-breaks-sid-playback

link is inaccessible.

Is there a specific version/commit that introduced the problem?
Can't reproduce here on Linux/x86-64, maybe a platform/compiler or optimization level issue?
In fact it sounds like an uninitialized var, can you try a run with compile sanitizers (UBSan/ASan)?


Tunes path for future reference:
`/MUSICIANS/T/Tel_Jeroen/RoboCop_3.sid`
`/MUSICIANS/E/Eco/Love_Is_a_Shield.sid`

**Comment by vitamin-caig** [comment]
> link is inaccessible.

Looks like bitbucket faced some issues... MUSICIANS/P/PAGE_JASON/Sixth.sid mentioned as problematic

> Is there a specific version/commit that introduced the problem?

No, I've tried vanilla 2.7.1 (before my changes) - problem was actual. Switching to old filter solveв problem according to my and another reports.

> In fact it sounds like an uninitialized var, can you try a run with compile sanitizers (UBSan/ASan)?

No luck...

**Comment by vitamin-caig** [comment]
Gotcha! UBsan gave that:

```
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:877:51: runtime error: index -1 out of bounds for type 'short unsigned int [65536]'    
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:1866:43: runtime error: index -942 out of bounds for type 'short unsigned int [65536]' 
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:1866:43: runtime error: index 65996 out of bounds for type 'short unsigned int [65536]'
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:1863:6: runtime error: signed integer overflow: 2038571051 - -160565658 cannot be represented in type 'int'
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:1863:6: runtime error: signed integer overflow: -2049526138 - 105802092 cannot be represented in type 'int'
 sidplayfp/src/builders/resid-builder/resid/filter8580new.h:877:63: runtime error: index -14034 out of bounds for type 'short unsigned int [1310720]'
```

**Comment by drfiemost** [comment]
Are you changing the filter bias value? I can't reproduce with the default setting.

**Comment by vitamin-caig** [comment]
No. But filter is enabled and fast sampling is used with interpolation.

**Comment by drfiemost** [comment]
With

`src/sidplayfp --resid -rif /MUSICIANS/T/Tel_Jeroen/RoboCop_3.sid`

I get
`../src/builders/resid-builder/resid/extfilt.h:146:50: runtime error: left shift of negative value -2514`

😕

**Comment by vitamin-caig** [comment]
Maybe also you need powerOnDelay = 8190

My report is from Love_Is_A_Shield.sid.
From Robocop3:
```
sidplayfp/src/builders/resid-builder/resid/filter8580new.h:877:51: runtime error: index 65536 out of bounds for type 'short unsigned int [65536]'
sidplayfp/src/builders/resid-builder/resid/filter8580new.h:1866:43: runtime error: index 65594 out of bounds for type 'short unsigned int [65536]'
```

I will try to collect all reports from HVSC later.

You may try to build zxtune on commit bcc335106440. (https://github.com/vitamin-caig/zxtune.git)

variables.mak at the root:
```
platform=linux
arch=x86_64
linux.cxx.flags=-fPIC -fsanitize=undefined                                                                                                                                                                        libraries.linux+=ubsan
```

make -C apps/zxtune123 -j <num_of_cpus>

**Comment by vitamin-caig** [comment]
Looks like I've found similar issue for old filter. Dependency chain is:
Vw_bias (unitialized) -> Vddt_Vw_2 (in set_w0) -> read from vcr_kVg in solve_integrate_*

**Comment by drfiemost** [comment]
> Looks like I've found similar issue for old filter. It's actual if ReSIDBuilder::bias is not called. Dependency chain is: Vw_bias (unitialized) -> Vddt_Vw_2 (in set_w0) -> read from vcr_kVg in solve_integrate_*

OK thanks, I'll report it upstream.
